userland/libutest/utest.c

```diff
@@ -0,0 +1,393 @@
+ * Remove a utest_t from the front of the linked list.
+ */
+static void list_pop(linked_list_t *list) {
+    if (!list->head) return;
+
+    list->head = list->head->next;
+
+    if (!list->head) list->tail = NULL;
+}
+
+/**
+ * Returns true if the linked list contains the given utest_t.
+ */
+static bool list_contains(linked_list_t *list, utest_t *test) {
+    utest_t *curr = list->head;
+    while (curr) {
+        if (curr == test) return true;
+    }
+    return false;
+}
+
+/*******************************************************************************
+ * GLOBAL STATE VARIABLES
+ ******************************************************************************/
+
+/**
+ * Test runner's shared buffer which holds the per-test-runner utest_t state.
+ * This must be aligned because the test runners share their buffers with the
+ * test supervisor via the `ipc_share` mechanism.
+ */
+#define TEST_BUF_SZ 128
+static char test_buf[TEST_BUF_SZ] __attribute__((aligned(TEST_BUF_SZ)));
+
+/**
+ * Test runner's condition variable which allows the test runner to
+ * cooperatively multiprogram with the test supervisor.
+ */
+static bool done = false;
+
+/**
+ * Test supervisor's linked list of pending test runners.
+ */
+static linked_list_t pending_pids;
+
+
+/*******************************************************************************
+ * TEST RUNNER FUNCTIONS
+ ******************************************************************************/
+
+/** \brief Set the `done` condition variable when notified by the supervisor.
+ *
+ * IPC callback which sets the `done` condition variable, allowing the test
+ * runner to pause execution and await supervisor approval to continue running.
+ */
+static void continue_cb(__attribute__ ((unused)) int pid,
+                       __attribute__ ((unused)) int arg2,
+                       __attribute__ ((unused)) int arg3,
+                       __attribute__ ((unused)) void *ud) {
+    done = true;
+}
+
+/** \brief Notify the test supervisor, and await approval to continue the test.
+ */
+static void sync_with_supervisor(int svc) {
+    done = false;
+    ipc_notify_svc(svc);
+    yield_for(&done);
+}
+
+/** \brief Run a sequence of unit tests and report the results.
+ *
+ * This function is called by the IPC clients, i.e. the 'test runners'. This
+ * function coordinates with the test supervisor's IPC service to run each test
+ * in sequence and report the status of each one.
+ *
+ * \param tests An array of boolean functions which return true for PASS and
+ *              false for FAIL.
+ * \param test_count The total number of tests in the tests array.
+ * \param timeout_ms The maximum amount of time each test is allowed to run
+ *                   before being timed out.
+ * \param svc_name The IPC service name of the test supervisor (e.g.
+ *                 "org.tockos.utest")
+ */
+void utest_runner(utest_fun *tests, uint32_t test_count,
+                  uint32_t timeout_ms, const char *svc_name) {
```

Comment:
This should have a return code so that the app can see the error where the supervisor service is not installed on the board.